KV client ADR

ripienaar opened the pull request:
This is an initial ADR that documents the design of a Key-Value clients,
the intention is that our main client languages will implement this in a
interoperable way.

This documents a reference implementation that is available in the nats
cli and under jsm.go

Signed-off-by: R.I.Pienaar <rip@devco.net>


/cc @nats-io/core

ripienaar commented on doc/adr/0008-jetstream-kv.md:
```diff
@@ -0,0 +1,326 @@
+|Date    |2021-06-30|
+|Author  |@ripienaar|
+|Status  |Partially Implemented|
+|Tags    |jetstream, client, kv|
```
@scottf here is an example of how I suggest we change the top-of-adr, we can then parse these tags and provide a `client` only list of associated ADRs etc, giving us the ability to have many views into the growing set of ADRs

We can list unapproved ones, approved ones, ones by impact etc.

ripienaar replied:
Here's an example of the Python PEP process.  They have 1 PEP list, PEP-1, PEP-2 etc much like the ADR process.

What they then have is views/indexes over these for like "Accepted PEPs" and so forth.

https://www.python.org/dev/peps/

The intention with the tags here is to have generated indexes for all ADRs that might need client implementations, or ones thats just kv related, or jetstream related - or even an intersection of the lot so one ADR will show up in 3 indexes in this case.

I don't think we can solve this with a file naming convention, so interested in views from the reviewers here, if we think this is a useful thing I can update the other ADRs at the same time to have this format

derekcollison commented on doc/adr/0008-jetstream-kv.md:
```diff
@@ -0,0 +1,326 @@
+Initial feature list:
+
+ * Multiple named buckets full of keys with n historical values kept per value
+ * Put and Get of `string(k)=string(v)` values
```
I would make the value opaque - []byte

ripienaar replied:
however the language implementation like, if its bytes or string, whatever works.

derekcollison commented on doc/adr/0008-jetstream-kv.md:
```diff
@@ -0,0 +1,326 @@
+ * Keys can be expired from the bucket based on a TTL
+ * Watching a specific key or the entire bucket for live updates
+ * Encoders and Decoders that transforms both keys and values
+ * A read-cache that builds an in-memory cache for fast reads
```
Be great if this could be intelligent and understand your working set for keys and read vs write etc.

ripienaar replied:
it pulls in the entire bucket at the moment, so probably not great for massive buckets with thousands of keys and huge values.  But it's a good start and other kinds of cache can be added.

The cache design is documented later on the doc, will work well for typical apps and support read replicas etc

derekcollison replied:
Feels like an impl detail but one that I think is important to discuss.

derekcollison commented on doc/adr/0008-jetstream-kv.md:
```diff
@@ -0,0 +1,326 @@
+	Close() error
+
+	// JSON dumps the entire KV as k=v values in JSON format
+	JSON(ctx context.Context) ([]byte, error)
```
Need to think through this one more ;)

ripienaar replied:
unfortunately given how this works, its what we need - its a push consumer and if ctx timeout before we reach end we need to bail out.

anyway, as with `[]byte` whatever makes a specific language happy as long as you make the bucket available - for the CLI this works for me.

derekcollison replied:
Yea but this is too simplistic and only works for small data sets. Need to think more about this IMO.

tpihl replied:
Do we need to limit value to string?  Nats allow slice of bytes for Data and I think that’s needed for a kv store as well.

ripienaar replied:
@tpihl As mentioned previously, languges can implement this however they like - of course I convert string to byte here because the reference implementation is there to support the CLI.

Language authors can do the right thing for their language.  I am not going to dictate to C/C++/Rust/Java/.Net to use a thing, they know what makes sense for their case.  This is a guide. Langauge implementers need to make their appropriate choices.

I was borrowing many concepts and ideas from etcd for example, their go API has 

```
ctx, cancel := context.WithTimeout(context.Background(), timeout)
resp, err := kvc.Put(ctx, "sample_key", "sample_value")
```

ripienaar commented on doc/adr/0008-jetstream-kv.md:
```diff
@@ -0,0 +1,326 @@
+ * Put and Get of `string(k)=string(v)` values
+ * Put only if the sequence of the last value for a key matches an expected sequence
+ * Historic values can be kept per key, setable on creation of the bucket
+ * Key deletes preserves history
```
Instead of deleting all messages for KV.BUCKET.x we put a message in saying its a delete operation.  Historical values are kept, client will return key not found on Get() if the last value for that subject is a delete op.

derekcollison replied:
Why would you want that? Just curious.

ripienaar replied:
A if I tell you I keep 5 historic values, then I should keep 5 historic values when if its 4 old + delete.  I think its important to be able to undo a delete.

On a more practical level, if I delete a subject with our new capabilities anyone consuming that stream - perhaps on a filter subjet - has no idea this happened.  This present a problem for things like the cache, because it will never evict a deleted key from the cache.

So this solves the cache consistency issue - but more importantly I think it preserved the idea behind historical values.

ripienaar commented on doc/adr/0008-jetstream-kv.md:
```diff
@@ -0,0 +1,326 @@
+ * Encoders and Decoders that transforms both keys and values
+ * A read-cache that builds an in-memory cache for fast reads
+ * read-after-write safety unless read replicas used
+ * Valid keys are `\A[-/_a-zA-Z0-9]+\z` after encoding
```
Yes, I like known limits though - we have this sliding scale even in the protocol spec that say "alpha numeric...but really anything and compatibility will depend on language" this does not guarantee interoperability.

Other annoying things like every week I see at least one person asking why they cant interact with our `$XXX` subjects from the CLI - they dont understand quoting, same problem with `!` and backticks etc.

Excluding some special characters isnt making this any worse imo

ripienaar commented on doc/adr/0008-jetstream-kv.md:
```diff
@@ -0,0 +1,326 @@
+	// Key is the key that was retrieved
+	Key() string
+	// Value is the retrieved value
+	Value() string
```
can be whatever the language wants as stated - if the nats.go implementation likes []byte go for it.  The interface here is basically a rough intent rather than the rule of the law. 

As long as the things are interpreted in the end `foo` as `foo` its fine.

ripienaar commented on doc/adr/0008-jetstream-kv.md:
```diff
@@ -0,0 +1,326 @@
+	// Sequence is a unique sequence for this value
+	Sequence() uint64
+	// Delta is distance from the latest value. If history is enabled this is effectively the index of the historical value, 0 for latest, 1 for most recent etc.
+	Delta() uint64
```
Delta is important - its an indication if this is the most recent value for a subject or bucket.

derekcollison replied:
But you really don't know once you received the reply and it will not age well if you hold onto it. What are we trying to achieve here?

derekcollison replied:
But you really don't know once you received the reply and it will not age well if you hold onto it. What are we trying to achieve here?

ripienaar replied:
A few things:

 * When building a cache from messages received from a watcher, I do not want the cache to be ready till it got the last message - else its giving old messages
 * When walking history of a key, I know when I am at the end.  Remember I am not simply passing the nats.Msg to users of a watch its a representation of it. Because later we want to support non nats.Msg based backends (NGS)

So this takes the important information we need and hides the per-implimentation detail.

ripienaar commented on doc/adr/0008-jetstream-kv.md:
```diff
@@ -0,0 +1,326 @@
+	// OriginCluster is the cluster where this data originate from, may be empty if sharing was not enabled
+	OriginCluster() string
+	// Operation is the kind of operation this result represents, enum of PUT and DEL
+	Operation() Operation
```
This is so that callers (like those on the other side of a watch) can know they are looking at a delete.

ripienaar commented on doc/adr/0008-jetstream-kv.md:
```diff
@@ -0,0 +1,326 @@
+	Values() uint64
+
+	// History returns the configured history kept per key
+	History() int64
```
can be whatever makes sense, but yet probably uint is best - but as []byte this isnt a stated law saying `nats.go` must match this :)

ripienaar commented on doc/adr/0008-jetstream-kv.md:
```diff
@@ -0,0 +1,326 @@
+	Cluster() string
+
+	// Replicas returns how many times data in the bucket is replicated at storage
+	Replicas() (ok int, failed int)
```
need a way to indicate good/bad replicas, probably this bit needs a rethink - I want to be sure not to force something like a NGS backend thats not JS based backend to contort themselves weirdly

ripienaar replied:
OK - replicas current and not lagged, failed - not active or very lagged.

ripienaar commented on doc/adr/0008-jetstream-kv.md:
```diff
@@ -0,0 +1,326 @@
+	Keys() ([]string, error)
+
+	// BackingStore is a backend specific name for the underlying storage - eg. stream name
+	BackingStore() string
```
informational...you migth not, or you might seems harmless to expose

derekcollison replied:
Could be the store engine? So in future could say sled or redis etc?

ripienaar replied:
sure and sled might have a internal name.

Here for example, I know I am using a jetstream instance and maybe I want to backup the stream.  Hows a user to know, I dont want to tell them and expect them to remember that its `KV_BUCKETNAME`, no, I say is this a JS backend? yes, ok backup the stream from BackingStore().

aricart replied:
This should also not be a string, but deferring to the constants/types for the current storage `StorageType` - I think we need to align some of these fields to the JetStream types they telegraph, avoiding yet more types.

ripienaar replied:
The backingstore here would not be something like file - here it would be the stream name, or maybe the consult backend name or etcd endpoint url or whatever makes sense for the backend to give hints to what it is.

Perhaps this could be something like a url even - `jetstream://KV_BOB`  and `etcd://some.node:123` etc, this way the CLI can figure out how to do bakcups etc?

ripienaar commented on doc/adr/0008-jetstream-kv.md:
````diff
@@ -0,0 +1,326 @@
+This is a read-only KV store handle, every backend should implement a language equivalent interface
+
+```go
+// RoKV is a read-only interface to a single key-value store bucket
````
I disagree, it has to be specific because I dont want on a RO KV the `Put()` to be possible at all, thats what interfaces are for.

derekcollison replied:
That feels like a permission thing, not a backend impl thing. Feels like it might be over complicating things.

ripienaar replied:
Interfaces in go are permissions.

derekcollison replied:
They can be, but in our world subjects are and we should always rely on those for the last say so to speak, at least imo.

ripienaar replied:
As a programmer, when I configure the client package to be read only, returning a subset interface with only reading methods on it communicates the intent - and abilities even during my time in the IDE.  It informats auto completion and more.

Why do we have io.Reader and io.ReadCloser? dont we just need io.ReadCloser when you can just ignore the Close() method? Because the language design tells you, your IDE and the compiler, that you are never going to or have to close this thing.  Likewise returning a read only interface means given 500 compatible backends - none of them will inappropriately expose write methods.

But again - language client details, ignore this if you like

ripienaar replied:
I clarified in the push why I called out the Ro and normal versions of the interface.

ripienaar commented on doc/adr/0008-jetstream-kv.md:
```diff
@@ -0,0 +1,326 @@
+
+Initial feature list:
+
+ * Multiple named buckets full of keys with n historical values kept per value
```
We dont support something to limit the number of keys in JS.  Versions per key is the History mentioned

Max size per value, Max size per bucket good, will add.

derekcollison replied:
Should we add max subjects to a stream definition?

ripienaar replied:
Might be interesting - then I can also get from JS a list of subjects and count of subjects - these are why I dont have easy to access things like `Keys()` because I have to walk the entire stream to answer those questions.

derekcollison replied:
OK I think this might be needed, but we can do simple math in the meantime.

ripienaar replied:
@matthiashanel yes, it will be removing a message to make space so should be allowed to write to a full bucket.

ripienaar commented on doc/adr/0008-jetstream-kv.md:
```diff
@@ -0,0 +1,326 @@
+	TTL() time.Duration
+
+	// Cluster returns the name of the cluster holding the read replica of the data
+	Cluster() string
```
Of course it is :) I anticipate the kv client - long running ones - will load up configuration that says if I am connected to cluster `lon` then my read replica is `CFG_LON` or whatever. I have some design for this but not included here yet while I noodle it over.

So when looking at any given cluster full of read replicas and so forth and so on, being able to audit them and more it's important to know where they are hosted

ripienaar commented on doc/adr/0008-jetstream-kv.md:
```diff
@@ -0,0 +1,326 @@
+	Replicas() (ok int, failed int)
+
+	// Keys returns a list of all keys in the bucket - not possible now
+	Keys() ([]string, error)
```
something for nats.go to do in the same way you do other similar things.  For CLI this is fine.  Again this isnt the law for how nats.go should do it - its the intent, make language implementations as makes most sense.

aricart replied:
Keys should be part of the bucket API. Getting the status should be fast. Until supported by JS, it requires walking the entire bucket.

aricart commented:
my only comment is that I do think that the values should always be []byte. If clients choose to only do strings, some values will never be readable across clients. If the base signature for Get/Set is []byte, then cross-client deserialization can be what the application requires.

ripienaar commented:
> my only comment is that I do think that the values should always be []byte. If clients choose to only do strings some values may never be readable across clients. If the base signature for Get/Set is []byte, then cross-client deserialization can be what the application requires.

Does this not make it clear?

```
Below are some guidance around types, I think it's important that we all have a method called `Put()` but if your language
is predominantly `[]byte` based, then adjust to taste from the `string` shown here.

Some of these uses go `context.Context` other language might not have an equivalent and need to handle timeouts or cancellations in another way, thats fine, this is a rough guide, go ahead and adjust to what is idiomatic in your language.```

Obviously - as the data goes over NATS - its bytes underneath, I don't see how anything becomes impossible to read just because some language has a string in the API? How does ruby ever achieve anything then?

derekcollison commented:
Some comments.

1. Not sure we expose buckets here. Assuming a bucket is what is imported, so it matters for setup etc but not as part of API per se.
2. Result seems excessive. I just want the []byte with option to get more info if I need to. Also sequence may have no meaning with other BEs.
3. Status would remove cluster and possibly backing store.
4. Would prefer to rethink RoKV.


In general I think we should start with a much simpler API and make sure we are not exposing impl details in the API itself.

ripienaar commented:
> Some comments.
> 
>     1. Not sure we expose buckets here. Assuming a bucket is what is imported, so it matters for setup etc but not as part of API per se.

What do you mean not expose buckets? It's like a DB name or a stream, you need to have multiple bucket names.  Maybe your comment is about a specific part but I dont see where you mean that.

>     2. Result seems excessive. I just want the []byte with option to get more info if I need to. Also sequence may have no meaning with other BEs.

This is quite common in KVs, I've seen this in all the common ones I looked.  @aricart and I think we should not call it `Result` but the info here is important and needed for operations, I dont see a difference between typing `val` and `val.Value()` in real terms really. But I do see making it more work to get acess to metadata would be worse.

>     3. Status would remove cluster and possibly backing store.

Will remove the origin header, but i think backing store is needed (to know what to backup etc)

>     4. Would prefer to rethink RoKV.

languages can do whatever they want with that, so will leave it as a show of the concept.

> 
> In general I think we should start with a much simpler API and make sure we are not exposing impl details in the API itself.

ripienaar commented:
>     3. Status would remove cluster and possibly backing store.

For status I will keep cluster actually, this is important in managing read replicas. Not in the ADR but I will let you have per cluster read replicas and for auditing/managing etc we need to know where the specific bucket is hosted.

ripienaar commented:
I did:

 * Rename `Result` to `Entry`
 * Removed the origin cluster concept but kept report on where a bucket is hosted
 * Renamed `Cluster` in status to `BucketLocation` which is more generic and less implementation specific

matthiashanel commented on doc/adr/0008-jetstream-kv.md:
```diff
@@ -0,0 +1,319 @@
+
+## Storage Backends
+
+We do have plans to support, and provide, commercial KV as part of our NGS offering, however there will be value in an
+open source KV implementation that can operate outside of NGS, especially one with an identical API.
```
Because of this, add Tyler to the review?

matthiashanel commented on doc/adr/0008-jetstream-kv.md:
```diff
@@ -0,0 +1,319 @@
+ * Historic values can be kept per key, setable on creation of the bucket
+ * Key deletes preserves history
+ * Keys can be expired from the bucket based on a TTL
+ * Watching a specific key or the entire bucket for live updates
```
Will this allow me to iterate over all present values in the order they where received?

aricart replied:
The interesting operation that is not currently possible is to get all the last values. - ie, per last message per subject.

matthiashanel replied:
Since messages are received in order, the last value will be received after it's corresponding historical values. 
I'm looking at it from the perspective of learning about all keys, so for me that would work. 

Only retrieving the last value has some more implications. Say you got Key A through this, but before you are done iterating Key A changes. What happens then?

I wonder, if only getting the last value is important to you, would you even configure them?

aricart replied:
Of course, because think about how you can use that as a log - for example the historic changes to all JWTs for an account.

matthiashanel replied:
You got me convinced, skipping old keys would nice.

ripienaar replied:
@matthiashanel yeah the cache is built with a bucket watch - it starts at oldest message and gets every message and write/delete/write etc as the entries flow.  Once it reaches pending=0 it knows the entire cache is built and consisten and then it turns from a passthrough cache to a one thats serving results.

So reading entire history - and indeed discovering all keys etc - is an option, but we want to add a API for getting all known subjects to JS since it knows that now and can gie it quickly.

matthiashanel commented on doc/adr/0008-jetstream-kv.md:
```diff
@@ -0,0 +1,319 @@
+
+When a stream is empty at create time - `NumPending+Delivered.Consumer==0` - a nil is sent across the channel to signal to the
+caller that they should not expect any values.
+
```
Do values contain an indication to wether the current value is currently the last one int the stream?
smth like seqno==max_seqno
Knowing when to stop would allow Watch to be used to iterate.

aricart replied:
if you look at `Delta` - it has the last value on the key when 0 (this is on history). For watchers, a key watcher only gets the last value, and then notifications from each new modification. For bucket watchers, you get all values and all modifications (this is why a JetStream consumer option of LastValuePerSubject would be very useful).

matthiashanel replied:
I read the delta as per key. not as current last.

ripienaar replied:
The delta is per key yes.  We dont know the key delta on a bucket watch though. But in practise thats fine because the watch is used to build a consistent local view of the KV - so once I am at pending=0 for the bucket it implies delta for all keys are 0.

aricart commented:
The other needed feature from StreamOption is an option that is sensical for max number of messages on the stream on a per subject. IE. If you enable history per subject of 10, and have a limit of 100 messages on the stream, unless we have a smart option that reduces the subject with the most/oldest key to be evicted, there's no room for one more key. Or worse, a key with a single value could be evicted because it is very old - that sort of behaviour would be unexpected.

aricart reviewed (commented):
LGTM - I have a JavaScript version already in the works here:
https://github.com/nats-io/nats.deno/pull/172

aricart commented:
delta is per key, but can only be calculated when watching a single key
(currently this is based on the number of messages remaining for the
consumer).

On Thu, Jul 8, 2021 at 1:58 PM Matthias Hanel ***@***.***>
wrote:

> ***@***.**** commented on this pull request.
> ------------------------------
>
> In doc/adr/0008-jetstream-kv.md
> <https://github.com/nats-io/nats-server/pull/2328#discussion_r666445862>:
>
> > +
> +This is a consumer set up to get the last value for a key. The watch remains active till cancelled and any PUT/DEL on this key should be
> +made available to the client.
> +
> +##### Bucket Watch
> +
> +A bucket watch is a consumer set up on the wildcard subject for the bucket - `$KV.<Bucket Name>.*` - and sending every single message
> +over the channel to the client.
> +
> +This means historical values will be sent as well as latest values. In this case we cannot provide the delta on a per key basis. But
> +the delta will represent the remainder for the entire bucket, this allows Watcher clients like the memory cache to set themselves
> +ready once the full history was read.
> +
> +When a stream is empty at create time - `NumPending+Delivered.Consumer==0` - a nil is sent across the channel to signal to the
> +caller that they should not expect any values.
> +
>
> I read the delta as per key. not as current last.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/nats-io/nats-server/pull/2328#discussion_r666445862>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAH4GEAU36G544H4NL46QI3TWXYONANCNFSM47SKZA6Q>
> .
>

ripienaar commented on doc/adr/0008-jetstream-kv.md:
```diff
@@ -0,0 +1,319 @@
+ * Encoders and Decoders that transforms both keys and values
+ * A read-cache that builds an in-memory cache for fast reads
+ * read-after-write safety unless read replicas used
+ * Valid keys are `\A[-/_a-=zA-Z0-9]+\z` after encoding
```
subject limits for now, I guess there would need to be some per backend limits? not sure, happy to add a configurable later on if needed

derekcollison commented:
After chatting online with @ripienaar I understand a bit better the direction and we discussed my issues. So LGTM.

ripienaar commented:
> The other needed feature from StreamConfig is an option that is sensical for max number of messages on the stream on a per subject. IE. If you enable history per subject of 10, and have a limit of 100 messages on the stream, unless we have a smart option that reduces the subject with the most/oldest key to be evicted, there's no room for one more key. Or worse, a key with a single value could be evicted because it is very old - that sort of behaviour would be unexpected.

Thats what max per subjet is.  Every subject keeps 5 messages. You don't then set a message limit for the stream as a whole.

derekcollison reviewed (approved):
LGTM per our chat.

aricart commented on doc/adr/0008-jetstream-kv.md:
```diff
@@ -0,0 +1,319 @@
+// RoKV is a read-only interface to a single key-value store bucket
+type RoKV interface {
+	// Get gets a key from the store
+	Get(key string) (Entry, error)
```
so one thing to consider is that if a key doesn't exist, we are considering that an error (which yes as far as JetStream is concerned it is) but in applications, it is NOT an error. It simply means that the value is not there. Imho the return value for go should be (*Entry, error). I know for all other JSM/etc we simply return an error, but maybe that is OK there, because really it could be a number of other things. However in this case, not finding is OK, it is just not there.

aricart replied:
To illustrate:
```javascript
          try {
            const ba = await this.billing.get(ac.sub);
          } catch (err) {
            //have to handle not found as an error!
            planID = "free"
            console.log(err);
          }
          // OR 
          const ba = await this.billing.get(ac.sub);
          const plan = ba ? ba.plan : "free";
```

```go
           ba, err := bucket.get(ac.sub)
           if  err != nil {
             return err;
           }
           planID := "free"
           if ba != nil {
              planID = ba.Plan
           }
```

ripienaar replied:
Entry is an interface so can’t be *Entry but it is nil on not found so I think it’s the same.

ripienaar commented:
Merged into main ADR repo after review here. No doubt follow ups as we use/implment
